_posts/2024-09-05-digging-into-struct-initialization-performance.md

```diff
@@ -0,0 +1,251 @@
+
+```crystal
+require "benchmark"
+require "blake3"
+require "digest/sha256"
+
+struct Blake3Hasher
+  def initialize
+    @hasher = uninitialized UInt8[1912]
+    Digest::Blake3::Lib.init(self)
+  end
+
+  def update(data)
+    Digest::Blake3::Lib.update(self, data.to_slice, data.bytesize)
+  end
+
+  def final(hashsum)
+    Digest::Blake3::Lib.final(self, hashsum, hashsum.size)
+  end
+
+  def to_unsafe
+    pointerof(@hasher)
+  end
+end
+
+def blake3_hexstring(data)
+  hasher = Blake3Hasher.new
+  hashsum = uninitialized UInt8[32]
+  hasher.update(data)
+  hasher.final(hashsum.to_slice)
+  hashsum.to_slice.hexstring
+end
+
+Benchmark.ips do |x|
+  bytes = Random::Secure.random_bytes(32)
+  x.report("SHA256") { Digest::SHA256.hexdigest(bytes) }
+  x.report("Blake3") { blake3_hexstring(bytes) }
+end
+```
+
+```
+SHA256   1.34M (744.61ns) (± 0.52%)   225B/op   1.38× slower
+Blake3   1.85M (539.81ns) (± 1.22%)  80.0B/op        fastest
+```
+
+And… performance is crashing down.
+
+## What is going on?
+
+We still only allocate 80B in the HEAP because structs are allocated on the stack, so what’s going on? Aren’t wrapping structs supposed to be free abstractions in Crystal? Because this is clearly not the case here. Is LLVM failing to inline the struct and method calls, and is BLAKE3 so fast that any overhead degrades performance?
+
+There are no more hints to understand what’s happening here. We must dig into the generated code to do any further investigation. I generated the LLVM IR for the above benchmark: 
+
+```
+crystal build --release --emit llvm-ir --no-debug bench.cr
+```
+
+The above command will generate a `bench.ll` file in the current directory. We build in release mode so that LLVM will optimize the code, and tell Crystal to not generate debug information to improve the readability. Sadly, I couldn't find anything weird there.
+
+> **NOTE:**
+> While I’m writing this blog post I notice that the LLVM IR does exhibit the exact same issue as the disassembly below. I have no idea how I didn’t notice it… maybe I forgot `--release` 🤦
+
+Let’s go deeper and inspect the generated assembly for my CPU. We can disassemble the executable with `objdump -d` for example, and this time I immediately noticed something weird: the function calls `Blake3::Hasher.new` followed by pages of MOV instructions, while the disassembly of the direct C function calls are just a few instructions and function calls.
+
+Here is the disassembly for the slow `struct` case. Looking into the rest of the disassembly, the same is happening inside `Blake3Hasher.new` when initializing `@hasher`: too many repeated MOV instructions to count.
+
+```
```

Comment:
**issue:** This also needs an explicit language tag. Maybe something in https://rouge-ruby.github.io/docs/file.Languages.html will fit for this? For example `armasm` seems nice - despite not being arm assembly. Otherwise `text` should do.